be/src/exec/hash_join_components.cpp

```diff
@@ -155,4 +385,380 @@ ChunkPtr SingleHashJoinBuilder::convert_to_spill_schema(const ChunkPtr& chunk) c
+    _partition_join_max_rows = 0;
+    _probe_estimated_costs = 0;
+    _fit_L2_cache_max_rows = 0;
+    _fit_L3_cache_max_rows = 0;
+    _pushed_chunks = 0;
+}
+
+void AdaptivePartitionHashJoinBuilder::reset(const HashTableParam& param) {
+    close();
+    create(param);
+}
+
+bool AdaptivePartitionHashJoinBuilder::anti_join_key_column_has_null() const {
+    return std::any_of(_builders.begin(), _builders.end(),
+                       [](const auto& builder) { return builder->anti_join_key_column_has_null(); });
+}
+
+void AdaptivePartitionHashJoinBuilder::get_build_info(size_t* bucket_size, float* avg_keys_per_bucket) {
+    size_t total_bucket_size = 0;
+    float total_keys_per_bucket = 0;
+    for (const auto& builder : _builders) {
+        size_t bucket_size = 0;
+        float keys_per_bucket = 0;
+        builder->get_build_info(&bucket_size, &keys_per_bucket);
+        total_bucket_size += bucket_size;
+        total_keys_per_bucket += keys_per_bucket;
+    }
+    *bucket_size = total_bucket_size;
+    *avg_keys_per_bucket = total_keys_per_bucket / _builders.size();
+}
+
+size_t AdaptivePartitionHashJoinBuilder::get_output_probe_column_count() const {
+    return _builders[0]->get_output_probe_column_count();
+}
+
+size_t AdaptivePartitionHashJoinBuilder::get_output_build_column_count() const {
+    return _builders[0]->get_output_build_column_count();
+}
+
+int64_t AdaptivePartitionHashJoinBuilder::ht_mem_usage() const {
+    return std::accumulate(_builders.begin(), _builders.end(), 0L,
+                           [](int64_t sum, const auto& builder) { return sum + builder->ht_mem_usage(); });
+}
+
+Status AdaptivePartitionHashJoinBuilder::_convert_to_single_partition() {
+    // merge all partition data to the first partition
+    for (size_t i = 1; i < _partition_num; ++i) {
+        _builders[0]->hash_table().merge_ht(_builders[i]->hash_table());
+    }
+    _builders.resize(1);
+    _partition_num = 1;
+    return Status::OK();
+}
+
+Status AdaptivePartitionHashJoinBuilder::_append_chunk_to_partitions(const ChunkPtr& chunk) {
+    const std::vector<ExprContext*>& build_partition_keys = _hash_joiner.build_expr_ctxs();
+
+    size_t num_rows = chunk->num_rows();
+    size_t num_partitions = _builders.size();
+    size_t num_partition_cols = build_partition_keys.size();
+
+    std::vector<ColumnPtr> partition_columns(num_partition_cols);
+    for (size_t i = 0; i < num_partition_cols; ++i) {
+        ASSIGN_OR_RETURN(partition_columns[i], build_partition_keys[i]->evaluate(chunk.get()));
+    }
+    std::vector<uint32_t> hash_values;
+    {
+        hash_values.assign(num_rows, HashUtil::FNV_SEED);
+
+        for (const ColumnPtr& column : partition_columns) {
+            column->fnv_hash(hash_values.data(), 0, num_rows);
+        }
+        // find partition id
+        for (size_t i = 0; i < hash_values.size(); ++i) {
+            hash_values[i] = HashUtil::fmix32(hash_values[i]) & (num_partitions - 1);
```

Comment:
why use  HashUtil::fmix32(hash_values[i]) & (num_partitions - 1) instead of 
hash_values[i] % (num_partitions-1)?  num_partitions is always power of 2?

Reply:
1. our shuffle hash function is fnv hash. We use an additional rehash function to make the hash value more hashed.
2. num_partitions is always power of 2